File: ios/Ubapp/Games/Tag/BleProximity.swift
import CoreBluetooth
import Foundation

/// Combined BLE central (scan) + peripheral (advertise) for one tag round.
/// Your phone sees other Ubapp peers AND advertises so others see you.
///
/// Peer id is carried in the advertisement's local name (kCBAdvDataLocalName)
/// and verified against [kUbappTagServiceUuid] on the scan side. iOS limits
/// advertising in the background; this works fine while the app is foreground.
final class BleProximityRuntime: NSObject, ProximitySource {
    enum AdvertiseStatus { case idle, starting, advertising, stopped, error, unavailable }

    let selfPeerId: String
    private let serviceUuid: CBUUID

    var onEvent: ((ProximityEvent) -> Void)?
    var onAdvertiseStatus: ((AdvertiseStatus, String?) -> Void)?

    private var central: CBCentralManager?
    private var peripheral: CBPeripheralManager?
    private var wantsScan = false
    private var wantsAdvertise = false

    init(selfPeerId: String, serviceUuid: String = kUbappTagServiceUuid) {
        self.selfPeerId = selfPeerId
        self.serviceUuid = CBUUID(string: serviceUuid)
        super.init()
    }

    func start() {
        wantsScan = true; wantsAdvertise = true
        if central == nil {
            central = CBCentralManager(delegate: self, queue: .main,
                options: [CBCentralManagerOptionShowPowerAlertKey: false])
        }
        if peripheral == nil {
            peripheral = CBPeripheralManager(delegate: self, queue: .main,
                options: [CBPeripheralManagerOptionShowPowerAlertKey: false])
        }
        tryScan(); tryAdvertise()
    }

    func stop() {
        wantsScan = false; wantsAdvertise = false
        central?.stopScan()
        peripheral?.stopAdvertising()
        onAdvertiseStatus?(.stopped, nil)
    }

    private func tryScan() {
        guard wantsScan, let c = central, c.state == .poweredOn else { return }
        c.scanForPeripherals(withServices: [serviceUuid],
                             options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
    }

    private func tryAdvertise() {
        guard wantsAdvertise, let p = peripheral, p.state == .poweredOn else { return }
        onAdvertiseStatus?(.starting, nil)
        p.startAdvertising([
            CBAdvertisementDataServiceUUIDsKey: [serviceUuid],
            CBAdvertisementDataLocalNameKey: selfPeerId,
        ])
    }
}

extension BleProximityRuntime: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn: tryScan()
        case .unsupported, .unauthorized: onAdvertiseStatus?(.unavailable, "scan unavailable")
        default: break
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData: [String: Any], rssi RSSI: NSNumber) {
        // Peer id: prefer the advertisement's kCBAdvDataLocalName, fall back
        // to the peripheral's name.
        let advertisedName = advertisementData[CBAdvertisementDataLocalNameKey] as? String
        let peerId = (advertisedName ?? peripheral.name ?? "").trimmingCharacters(in: .whitespaces)
        guard !peerId.isEmpty else { return }
        onEvent?(ProximityEvent(
            peerId: peerId, rssi: RSSI.intValue,
            atMs: Int64(Date().timeIntervalSince1970 * 1000)))
    }
}

extension BleProximityRuntime: CBPeripheralManagerDelegate {
    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        switch peripheral.state {
        case .poweredOn: tryAdvertise()
        case .unsupported, .unauthorized: onAdvertiseStatus?(.unavailable, "advertise unavailable")
        default: onAdvertiseStatus?(.idle, nil)
        }
    }

    func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: (any Error)?) {
        if let error { onAdvertiseStatus?(.error, error.localizedDescription) }
        else { onAdvertiseStatus?(.advertising, nil) }
    }
}

File: android/app/src/main/java/com/example/jamboree/games/tag/BleProximity.kt
package com.example.ubapp.games.tag

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.ParcelUuid
import java.util.UUID

/**
 * Combined BLE central (scan) + peripheral (advertise) for one tag round.
 * Your phone sees other Ubapp peers AND advertises so others see you.
 *
 * Peer id is carried as a one-byte+UTF-8 service-data blob keyed by
 * [UBAPP_TAG_SERVICE_UUID]. On Android the adapter's name is also used as a
 * fallback for older OS versions where service data isn't honored.
 *
 * Callers must hold the runtime BLUETOOTH_SCAN / BLUETOOTH_ADVERTISE
 * permissions before calling [start] — the manifest already declares them.
 */
@SuppressLint("MissingPermission")
class BleProximityRuntime(
    context: Context,
    val selfPeerId: String,
    serviceUuid: String = UBAPP_TAG_SERVICE_UUID,
) : ProximitySource {
    enum class AdvertiseStatus { IDLE, STARTING, ADVERTISING, STOPPED, ERROR, UNAVAILABLE }

    override var onEvent: ((ProximityEvent) -> Unit)? = null
    var onAdvertiseStatus: ((AdvertiseStatus, String?) -> Unit)? = null

    private val serviceUuid = UUID.fromString(serviceUuid)
    private val parcelUuid = ParcelUuid(this.serviceUuid)
    private val manager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val adapter: BluetoothAdapter? = manager.adapter

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val rec = result.scanRecord ?: return
            val data = rec.serviceData?.get(parcelUuid)
            val peerId = data?.let { String(it).trim() } ?: rec.deviceName?.trim().orEmpty()
            if (peerId.isEmpty()) return
            onEvent?.invoke(ProximityEvent(peerId, result.rssi, System.currentTimeMillis()))
        }
        override fun onScanFailed(errorCode: Int) {
            onAdvertiseStatus?.invoke(AdvertiseStatus.ERROR, "scan failed: $errorCode")
        }
    }

    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
            onAdvertiseStatus?.invoke(AdvertiseStatus.ADVERTISING, null)
        }
        override fun onStartFailure(errorCode: Int) {
            onAdvertiseStatus?.invoke(AdvertiseStatus.ERROR, "advertise failed: $errorCode")
        }
    }

    override fun start() {
        val a = adapter ?: return onAdvertiseStatus?.invoke(AdvertiseStatus.UNAVAILABLE, "no BT adapter") ?: Unit
        if (!a.isEnabled) {
            onAdvertiseStatus?.invoke(AdvertiseStatus.UNAVAILABLE, "BT off"); return
        }

        a.bluetoothLeScanner?.startScan(
            listOf(ScanFilter.Builder().setServiceUuid(parcelUuid).build()),
            ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY).build(),
            scanCallback,
        )

        val advertiser = a.bluetoothLeAdvertiser
        if (advertiser == null) {
            onAdvertiseStatus?.invoke(AdvertiseStatus.UNAVAILABLE, "advertiser unsupported"); return
        }
        onAdvertiseStatus?.invoke(AdvertiseStatus.STARTING, null)
        advertiser.startAdvertising(
            AdvertiseSettings.Builder()
                .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY)
                .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_HIGH)
                .setConnectable(false)
                .build(),
            AdvertiseData.Builder()
                .setIncludeDeviceName(false)
                .addServiceUuid(parcelUuid)
                .addServiceData(parcelUuid, selfPeerId.toByteArray())
                .build(),
            advertiseCallback,
        )
    }

    override fun stop() {
        adapter?.bluetoothLeScanner?.stopScan(scanCallback)
        adapter?.bluetoothLeAdvertiser?.stopAdvertising(advertiseCallback)
        onAdvertiseStatus?.invoke(AdvertiseStatus.STOPPED, null)
    }
}
